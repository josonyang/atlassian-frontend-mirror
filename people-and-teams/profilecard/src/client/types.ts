export type DEPRECATED_ErrorAttributes = {
	errorCount?: number;
	errorDetails?: Omit<DEPRECATED_ErrorAttributes, 'errorDetails'>[];
	errorMessage: string;
	errorCategory?: string;
	errorType?: string;
	errorPath?: string;
	errorNumber?: number;
	isSLOFailure: boolean;
	traceId?: string | null;
	errorStatusCode?: number;
};

export type ErrorAttributes = {
	errorCount: number | null;
	errorDetails: Omit<ErrorAttributes, 'errorDetails'>[] | null;
	errorMessage: string;
	errorCategory: string | null;
	errorType: string | null;
	errorPath: string | null;
	errorNumber: number | null;
	isSLOFailure: boolean;
	traceId: string | null;
	errorStatusCode: number | null;
};
