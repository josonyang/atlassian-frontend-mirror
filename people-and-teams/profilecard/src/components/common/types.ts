import { type AnalyticsEventPayload } from '@atlaskit/analytics-next';
import { type PopupProps } from '@atlaskit/popup';

import { type ProfileCardErrorType } from '../../types';

export type ProfileCardTriggerProps<T> = {
	trigger: 'hover' | 'click';
	ariaLabelledBy?: string;
	disabledAriaAttributes?: boolean;
	children: React.ReactNode;
	renderProfileCard: ({
		profileData,
		error,
	}: {
		profileData?: T;
		error: ProfileCardErrorType | undefined | null;
	}) => React.ReactNode;
	fetchProfile?: () => Promise<T>;
	profileCardType: 'agent' | 'user' | 'team';
	fireAnalytics?: (payload: AnalyticsEventPayload) => void;
} & Omit<PopupProps, 'trigger' | 'isOpen' | 'content'>;
