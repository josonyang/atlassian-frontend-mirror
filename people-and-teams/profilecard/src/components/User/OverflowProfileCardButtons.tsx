import React, { useCallback, useState } from 'react';

import { useIntl } from 'react-intl-next';

import { IconButton } from '@atlaskit/button/new';
import DropdownMenu, { DropdownItem, DropdownItemGroup } from '@atlaskit/dropdown-menu';
import MoreIcon from '@atlaskit/icon/core/migration/show-more-horizontal--more';
import { fg } from '@atlaskit/platform-feature-flags';

import messages from '../../messages';
import { OverflowActionButtonsWrapper } from '../../styled/Card';
import { type AnalyticsWithDurationProps, type ProfileCardAction } from '../../types';
import { moreActionsClicked } from '../../util/analytics';

type OverflowButtonsProps = {
	actions: ProfileCardAction[];
	onItemClick: (
		action: ProfileCardAction,
		args: any,
		event: React.MouseEvent | React.KeyboardEvent,
		index: number,
	) => void;
	fullName?: string;
} & AnalyticsWithDurationProps;

export const ACTION_OVERFLOW_THRESHOLD = 2;

export const OverflowProfileCardButtons = (props: OverflowButtonsProps) => {
	const intl = useIntl();

	const [, setOpen] = useState(false);

	const { actions, onItemClick, fireAnalyticsWithDuration } = props;

	const numActions = actions.length + ACTION_OVERFLOW_THRESHOLD;

	const onOpenChange = useCallback(
		({ isOpen: nextOpen }: { isOpen: boolean }) => {
			setOpen((prevOpen) => {
				if (nextOpen && !prevOpen) {
					fireAnalyticsWithDuration((duration) =>
						moreActionsClicked('user', {
							duration,
							numActions,
						}),
					);
				}

				return nextOpen;
			});
		},
		[numActions, fireAnalyticsWithDuration],
	);

	return (
		<OverflowActionButtonsWrapper testId="profilecard-actions-overflow">
			<DropdownMenu
				onOpenChange={onOpenChange}
				placement={'bottom-end'}
				trigger={({ triggerRef, isSelected, testId, ...providedProps }) => {
					return (
						<IconButton
							type="button"
							{...providedProps}
							ref={triggerRef}
							label={
								fg('jfp_a11y_team_profile_card_actions_label')
									? intl.formatMessage(messages.profileCardMoreIconLabelWithName, {
											fullName: props.fullName,
										})
									: intl.formatMessage(messages.profileCardMoreIconLabel)
							}
							icon={MoreIcon}
						/>
					);
				}}
			>
				<DropdownItemGroup>
					{actions.map((action, index) => (
						<DropdownItem
							key={action.id}
							onClick={(event, ...args: any) => {
								onItemClick(action, args, event, index);
							}}
							href={action.link}
							target={action.target}
						>
							{action.label}
						</DropdownItem>
					))}
				</DropdownItemGroup>
			</DropdownMenu>
		</OverflowActionButtonsWrapper>
	);
};
