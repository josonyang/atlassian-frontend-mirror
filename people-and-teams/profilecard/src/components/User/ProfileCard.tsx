import React, { useCallback, useEffect, useMemo, useState } from 'react';

import { FormattedMessage } from 'react-intl-next';

import { type AnalyticsEventPayload, withAnalyticsEvents } from '@atlaskit/analytics-next';
import Avatar from '@atlaskit/avatar';
import { LinkButton } from '@atlaskit/button/new';
import FocusRing from '@atlaskit/focus-ring';
import { fg } from '@atlaskit/platform-feature-flags';
import Spinner from '@atlaskit/spinner';
import { N0 } from '@atlaskit/theme/colors';
import { token } from '@atlaskit/tokens';

import messages from '../../messages';
import {
	ActionButtonGroup,
	ActionsFlexSpacer,
	AnimatedKudosButton,
	AnimationWrapper,
	CardContainer,
	CardContent,
	KudosBlobAnimation,
	ProfileImage,
} from '../../styled/Card';
import { CardWrapper, SpinnerContainer } from '../../styled/UserTrigger';
import {
	type AnalyticsFromDuration,
	type AnalyticsProps,
	type AnalyticsWithDurationProps,
	type ProfileCardAction,
	type ProfilecardProps,
} from '../../types';
import { actionClicked, fireEvent, profileCardRendered } from '../../util/analytics';
import { isBasicClick } from '../../util/click';
import { getPageTime } from '../../util/performance';
import { ErrorMessage } from '../Error';

import {
	ACTION_OVERFLOW_THRESHOLD,
	OverflowProfileCardButtons,
} from './OverflowProfileCardButtons';
import { ProfileCardDetails } from './ProfileCardDetails';

const GIVE_KUDOS_ACTION_ID = 'give-kudos';

const useKudos = (
	cloudId?: string,
	userId?: string,
	teamCentralBaseUrl?: string,
	openKudosDrawer?: () => void,
) => {
	const kudosUrl = useMemo(() => {
		const recipientId = userId ? `&recipientId=${userId}` : '';
		const cloudIdParam = cloudId ? `&cloudId=${cloudId}` : '';

		return `${teamCentralBaseUrl || ''}/kudos/give?type=individual${recipientId}${cloudIdParam}`;
	}, [cloudId, teamCentralBaseUrl, userId]);

	const kudosButtonCallback = useCallback(() => {
		if (openKudosDrawer) {
			openKudosDrawer();
		} else {
			window.open(kudosUrl);
		}
	}, [kudosUrl, openKudosDrawer]);

	const kudosAction = useMemo(() => {
		return {
			label: <FormattedMessage {...messages.giveKudosButton} />,
			id: GIVE_KUDOS_ACTION_ID,
			callback: kudosButtonCallback,
			link: kudosUrl,
		};
	}, [kudosButtonCallback, kudosUrl]);

	return {
		kudosAction,
		kudosButtonCallback,
		kudosUrl,
	};
};

const Wrapper = (props: { children: React.ReactNode }) => (
	<CardWrapper testId="profilecard" role="dialog" labelledBy="profilecard-name-label">
		{props.children}
	</CardWrapper>
);

export const ProfilecardInternal = (props: ProfilecardProps & AnalyticsProps) => {
	const [openTime] = useState<number>(getPageTime());

	const { createAnalyticsEvent } = props;

	const fireAnalytics = useCallback(
		(payload: AnalyticsEventPayload) => {
			if (createAnalyticsEvent) {
				fireEvent(createAnalyticsEvent, payload);
			}
		},
		[createAnalyticsEvent],
	);

	const fireAnalyticsWithDuration = useCallback(
		(generator: AnalyticsFromDuration) => {
			const elapsed = getPageTime() - openTime;
			const event = generator(elapsed);
			fireAnalytics(event);
		},
		[fireAnalytics, openTime],
	);

	const { kudosAction } = useKudos(
		props.cloudId,
		props.userId,
		props.teamCentralBaseUrl,
		props.openKudosDrawer,
	);

	const { actions = [], isCurrentUser, isKudosEnabled, status = 'active' } = props;

	const realActions = useMemo(() => {
		if (isCurrentUser || !isKudosEnabled || status !== 'active') {
			return actions;
		}

		return actions.concat([kudosAction]);
	}, [actions, isCurrentUser, isKudosEnabled, kudosAction, status]);

	const { isLoading, fullName, hasError } = props;

	const canRender = !hasError && !isLoading && !!(fullName || status === 'closed');

	useEffect(() => {
		if (canRender) {
			fireAnalyticsWithDuration((duration) =>
				profileCardRendered('user', 'content', {
					duration,
					numActions: realActions.length,
				}),
			);
		}
	}, [canRender, fireAnalyticsWithDuration, realActions]);

	if (hasError) {
		return (
			<Wrapper>
				<ErrorMessage
					reload={props.clientFetchProfile}
					errorType={props.errorType || null}
					fireAnalytics={fireAnalytics}
				/>
			</Wrapper>
		);
	}

	if (isLoading) {
		return (
			<Wrapper>
				<LoadingView fireAnalyticsWithDuration={fireAnalyticsWithDuration} />
			</Wrapper>
		);
	}

	if (!canRender) {
		return null;
	}

	const isDisabledUser = status === 'inactive' || status === 'closed';

	return (
		<Wrapper>
			<CardContainer isDisabledUser={isDisabledUser} withoutElevation={props.withoutElevation}>
				<ProfileImage>
					<Avatar
						size="xlarge"
						src={status !== 'closed' ? props.avatarUrl : undefined}
						borderColor={token('elevation.shadow.overlay', N0)}
					/>
				</ProfileImage>
				<CardContent>
					<ProfileCardDetails
						{...props}
						status={status}
						fireAnalyticsWithDuration={fireAnalyticsWithDuration}
					/>
					{realActions && (
						<>
							<ActionsFlexSpacer />
							<Actions
								actions={realActions}
								fireAnalyticsWithDuration={fireAnalyticsWithDuration}
								isTriggeredUsingKeyboard={props.isTriggeredUsingKeyboard}
							/>
						</>
					)}
				</CardContent>
			</CardContainer>
		</Wrapper>
	);
};

interface ActionsProps extends AnalyticsWithDurationProps {
	actions: ProfileCardAction[];
	isTriggeredUsingKeyboard: boolean | undefined;
}

const Actions = ({
	actions,
	fireAnalyticsWithDuration,
	isTriggeredUsingKeyboard,
}: ActionsProps) => {
	const onActionClick = useCallback(
		(
			action: ProfileCardAction,
			args: any,
			event: React.MouseEvent | React.KeyboardEvent,
			index: number,
		) => {
			fireAnalyticsWithDuration((duration) =>
				actionClicked('user', {
					duration,
					hasHref: !!action.link,
					hasOnClick: !!action.callback,
					index,
					actionId: action.id || 'no-id-specified',
				}),
			);

			if (action.callback && isBasicClick(event)) {
				event.preventDefault();
				action.callback(event, ...args);
			}
		},
		[fireAnalyticsWithDuration],
	);

	if (!actions || actions.length === 0) {
		return null;
	}

	const regularActions = actions.slice(0, ACTION_OVERFLOW_THRESHOLD);
	const overflowActions =
		actions.length > ACTION_OVERFLOW_THRESHOLD
			? actions.slice(ACTION_OVERFLOW_THRESHOLD)
			: undefined;

	return (
		<ActionButtonGroup testId="profilecard-actions">
			{regularActions.map((action, index) => {
				const isKudos = action.id === GIVE_KUDOS_ACTION_ID;

				const button = (
					<FocusRing isInset key={`profile-card-action-focus-ring_${action.id || index}`}>
						<LinkButton
							appearance="default"
							key={action.id || index}
							onClick={(event: React.MouseEvent<HTMLElement>, ...args: any) =>
								onActionClick(action, args, event, index)
							}
							href={action.link || ''}
							autoFocus={index === 0 && isTriggeredUsingKeyboard}
							id={`action-button-${action.id}`}
							aria-labelledby={
								fg('enable_userprofilecard_arialabelfix')
									? `action-button-${action.id} profilecard-name-label`
									: ''
							}
						>
							{action.label}
							{isKudos && (
								<AnimationWrapper>
									<KudosBlobAnimation />
								</AnimationWrapper>
							)}
						</LinkButton>
					</FocusRing>
				);

				if (isKudos) {
					return (
						<AnimatedKudosButton key={`profile-card-action-kudos_${action.id || index}`}>
							{button}
						</AnimatedKudosButton>
					);
				}

				return button;
			})}
			{overflowActions && (
				<OverflowProfileCardButtons
					actions={overflowActions}
					fireAnalyticsWithDuration={fireAnalyticsWithDuration}
					onItemClick={(action, args, event, index) =>
						onActionClick(action, args, event, index + ACTION_OVERFLOW_THRESHOLD)
					}
				/>
			)}
		</ActionButtonGroup>
	);
};

const LoadingView = ({ fireAnalyticsWithDuration }: AnalyticsWithDurationProps) => {
	useEffect(() => {
		fireAnalyticsWithDuration((duration) => profileCardRendered('user', 'spinner', { duration }));
	}, [fireAnalyticsWithDuration]);

	return (
		<SpinnerContainer testId="profilecard-spinner-container">
			<Spinner />
		</SpinnerContainer>
	);
};

export default withAnalyticsEvents()(ProfilecardInternal);
