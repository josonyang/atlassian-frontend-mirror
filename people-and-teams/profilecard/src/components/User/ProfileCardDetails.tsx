import React from 'react';

import { cssMap, cx } from '@compiled/react';
import { FormattedMessage } from 'react-intl-next';

import Lozenge from '@atlaskit/lozenge';
import { fg } from '@atlaskit/platform-feature-flags';
import { Text } from '@atlaskit/primitives';
import { Box } from '@atlaskit/primitives/compiled';
import { token } from '@atlaskit/tokens';

import relativeDate from '../../internal/relative-date';
import messages from '../../messages';
import {
	AppTitleLabel,
	CustomLozengeContainer,
	DetailsGroup,
	DisabledInfo,
	JobTitleLabel,
	LozengeWrapper,
} from '../../styled/Card';
import {
	type AnalyticsWithDurationProps,
	type LozengeProps,
	type ProfilecardProps,
} from '../../types';
import { IconLabel } from '../Icon';

import ReportingLinesDetails from './ReportingLinesDetails';

const styles = cssMap({
	detailedListWrapper: {
		marginTop: token('space.0'),
		marginRight: token('space.0'),
		marginBottom: token('space.0'),
		marginLeft: token('space.0'),
		paddingTop: token('space.0'),
		paddingRight: token('space.0'),
		paddingBottom: token('space.0'),
		paddingLeft: token('space.0'),
	},
	fullNameLabel: {
		overflow: 'hidden',
		textOverflow: 'ellipsis',
		whiteSpace: 'nowrap',
		font: token('font.body.large'),
	},
	noMetaLabel: {
		// Using `&` twice to increase specificity
		// eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors, @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
		'&&': {
			marginTop: token('space.400'),
			marginBottom: token('space.150'),
		},
		marginRight: '0',
		marginLeft: '0',
	},
	noMetaLabelWithHighSpecificity: {
		// Using `&` and id attribute to increase specificity
		// eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors
		'&#profilecard-name-label': {
			marginTop: token('space.400'),
			marginBottom: token('space.150'),
		},
	},
	metaLabel: {
		// Using `&` twice to increase specificity
		// eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors, @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
		'&&': {
			marginTop: token('space.150'),
		},
		marginRight: '0',
		marginBottom: '0',
		marginLeft: '0',
	},
	metaLabelWithHighSpecificity: {
		// Using `&` and id attribute to increase specificity
		// eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors
		'&#profilecard-name-label': {
			marginTop: token('space.150'),
		},
		marginRight: '0',
		marginBottom: '0',
		marginLeft: '0',
	},
	disabledAccount: {
		color: token('color.text'),
	},
	activeAccount: {
		color: token('color.text.inverse'),
	},
});

const renderName = (nickname?: string, fullName?: string, meta?: string) => {
	if (!fullName && !nickname) {
		return null;
	}

	const isNicknameRedundant = !nickname || nickname === fullName;
	const shownNickname = ` (${nickname}) `;

	const displayName = isNicknameRedundant ? fullName : `${fullName}${shownNickname}`;

	return (
		<Box
			as="h2"
			xcss={cx(
				styles.fullNameLabel,
				styles.activeAccount,
				meta
					? fg('platform_editor_profilecard_style_fix')
						? styles.metaLabelWithHighSpecificity
						: styles.metaLabel
					: fg('platform_editor_profilecard_style_fix')
						? styles.noMetaLabelWithHighSpecificity
						: styles.noMetaLabel,
			)}
			testId="profilecard-name"
			id="profilecard-name-label"
		>
			{displayName}
		</Box>
	);
};

const disabledAccountDesc = (
	statusModifiedDate: number | null | undefined,
	disabledAccountMessage: React.ReactNode | undefined,
	status: 'closed' | 'inactive' = 'closed',
) => {
	// consumer does not want to use built-in message
	if (disabledAccountMessage) {
		return disabledAccountMessage;
	}

	const date = statusModifiedDate ? new Date(statusModifiedDate * 1000) : null;
	const relativeDateKey = relativeDate(date);

	const msgKey = (
		relativeDateKey
			? `${status}AccountDescMsgHasDate${relativeDateKey}`
			: `${status}AccountDescMsgNoDate`
	) as keyof typeof messages;

	const secondSentence = <FormattedMessage {...messages[msgKey]} />;

	return (
		<Text as="p" testId="profilecard-disabled-account">
			<FormattedMessage {...messages.generalDescMsgForDisabledUser} /> {secondSentence}
		</Text>
	);
};

const CustomLozenges = ({ lozenges = [] }: { lozenges?: LozengeProps[] }) => {
	if (lozenges.length === 0) {
		return null;
	}

	return (
		<CustomLozengeContainer>
			{lozenges.map(({ text, ...otherProps }, index) => (
				<Lozenge {...otherProps} key={index}>
					{text}
				</Lozenge>
			))}
		</CustomLozengeContainer>
	);
};
const ServiceAccountProfileCardDetails = (props: ProfilecardProps) => {
	const { fullName, nickname } = props;

	return (
		<DetailsGroup>
			{renderName(nickname, fullName)}
			<AppTitleLabel>SERVICE ACCOUNT</AppTitleLabel>
		</DetailsGroup>
	);
};

const BotProfileCardDetails = (props: ProfilecardProps) => {
	const { fullName, nickname } = props;

	return (
		<DetailsGroup>
			{renderName(nickname, fullName)}
			<AppTitleLabel>APP</AppTitleLabel>
		</DetailsGroup>
	);
};

const DisabledProfileCardDetails = (
	props: ProfilecardProps & { status: 'closed' | 'inactive' },
) => {
	const {
		companyName,
		disabledAccountMessage,
		fullName,
		hasDisabledAccountLozenge = true,
		nickname,
		status,
		statusModifiedDate,
	} = props;

	const name =
		status === 'inactive'
			? fullName || nickname
			: nickname || <FormattedMessage {...messages.disabledAccountDefaultName} />;

	return (
		<DetailsGroup>
			<Box
				as="h2"
				xcss={cx(styles.fullNameLabel, styles.noMetaLabel, styles.disabledAccount)}
				testId="profilecard-name"
				id="profilecard-name-label"
			>
				{name}
			</Box>

			{hasDisabledAccountLozenge && (
				<LozengeWrapper>
					<Lozenge appearance="default" isBold>
						{status === 'inactive' ? (
							<FormattedMessage {...messages.inactiveAccountMsg} />
						) : (
							<FormattedMessage {...messages.closedAccountMsg} />
						)}
					</Lozenge>
				</LozengeWrapper>
			)}

			<DisabledInfo>
				{disabledAccountDesc(statusModifiedDate, disabledAccountMessage, status)}
			</DisabledInfo>

			{status === 'inactive' && <IconLabel icon="companyName">{companyName}</IconLabel>}
		</DetailsGroup>
	);
};

export const ProfileCardDetails = (props: ProfilecardProps & AnalyticsWithDurationProps) => {
	const { meta, status } = props;

	if (props.isServiceAccount) {
		return <ServiceAccountProfileCardDetails {...props} />;
	}

	if (props.isBot) {
		return <BotProfileCardDetails {...props} />;
	}

	if (status === 'inactive' || status === 'closed') {
		return <DisabledProfileCardDetails {...props} status={status} />;
	}

	return (
		<DetailsGroup>
			{renderName(props.nickname, props.fullName, meta)}
			{meta && <JobTitleLabel>{meta}</JobTitleLabel>}
			<CustomLozenges lozenges={props.customLozenges} />
			<Box as="dl" xcss={styles.detailedListWrapper}>
				<IconLabel icon="email" extraTopSpace={true}>
					{props.email}
				</IconLabel>
				<IconLabel icon="time">{props.timestring}</IconLabel>
				<IconLabel icon="companyName">{props.companyName}</IconLabel>
				<IconLabel icon="location">{props.location}</IconLabel>
			</Box>

			<ReportingLinesDetails
				reportingLines={props.reportingLines}
				reportingLinesProfileUrl={props.reportingLinesProfileUrl}
				onReportingLinesClick={props.onReportingLinesClick}
				fireAnalyticsWithDuration={props.fireAnalyticsWithDuration}
			/>
		</DetailsGroup>
	);
};
