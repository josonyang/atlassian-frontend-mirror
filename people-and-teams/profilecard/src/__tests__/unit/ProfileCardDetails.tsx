import React from 'react';

import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl-next';

import { fg } from '@atlaskit/platform-feature-flags';

import { ProfileCardDetails } from '../../components/User/ProfileCardDetails';
import { type LozengeProps } from '../../types';

jest.mock('react-intl-next', () => {
	return {
		...(jest.requireActual('react-intl-next') as any),
		useIntl: jest.fn().mockReturnValue({
			locale: 'en',
			formatMessage: (descriptor: any) => descriptor.defaultMessage,
			FormattedMessage: ({ descriptor }: any) => descriptor.defaultMessage,
		}),
	};
});

jest.mock('@atlaskit/platform-feature-flags', () => ({
	...jest.requireActual<any>('@atlaskit/platform-feature-flags'),
	fg: jest.fn(),
}));

type Props = Parameters<typeof ProfileCardDetails>[0];

const defaultProps: Props = {
	fireAnalyticsWithDuration: jest.fn(),
	fullName: 'full name test',
	status: 'active',
	nickname: 'jscrazy',
	companyName: 'Atlassian',
};

const renderComponent = (props: Partial<Props> = {}) =>
	render(
		<IntlProvider locale="en" defaultLocale="en-US">
			<ProfileCardDetails {...defaultProps} {...props} />
		</IntlProvider>,
	);

describe('ProfileCardDetails', () => {
	describe('name', () => {
		describe.each([
			['active user', false],
			['bot account', true],
		])('for %s', (_, isBot) => {
			it('should include nickname if applicable', () => {
				const { getByTestId } = renderComponent({
					isBot,
				});
				const component = getByTestId('profilecard-name');
				expect(component.textContent).toMatchInlineSnapshot(`"full name test (jscrazy) "`);
			});

			it('should only show name if no nickname', () => {
				const { getByTestId } = renderComponent({
					isBot,
					nickname: undefined,
				});
				const component = getByTestId('profilecard-name');
				expect(component.textContent).toMatchInlineSnapshot(`"full name test"`);
			});

			it('should show only once if nickname and full name match', () => {
				const { getByTestId } = renderComponent({
					isBot,
					fullName: 'Same name',
					nickname: 'Same name',
				});
				const component = getByTestId('profilecard-name');
				expect(component.textContent).toMatchInlineSnapshot(`"Same name"`);
			});

			it('should show nothing if no name or nickname provided', () => {
				const { queryByTestId } = renderComponent({
					isBot,
					fullName: undefined,
					nickname: undefined,
				});
				const component = queryByTestId('profilecard-name');
				expect(component).toBeNull();
			});

			it('should render name as a heading', () => {
				(fg as jest.Mock).mockReturnValue(true);
				const { getByRole } = renderComponent({
					isBot,
					fullName: 'Same name',
					nickname: 'Same name',
				});
				const component = getByRole('heading', { level: 2 });
				expect(component).toBeVisible();
				expect(component).toHaveTextContent('Same name');
			});
		});

		describe('for inactive user', () => {
			it('should show full name if available', () => {
				const { getByTestId } = renderComponent({
					status: 'inactive',
				});
				const component = getByTestId('profilecard-name');
				expect(component.textContent).toMatchInlineSnapshot(`"full name test"`);
			});

			it('should show nickname if no full name', () => {
				const { getByTestId } = renderComponent({
					status: 'inactive',
					fullName: undefined,
				});
				const component = getByTestId('profilecard-name');
				expect(component.textContent).toMatchInlineSnapshot(`"jscrazy"`);
			});
		});

		describe('for deleted user', () => {
			it('should show nickname if available', () => {
				const { getByTestId } = renderComponent({
					status: 'closed',
				});
				const component = getByTestId('profilecard-name');
				expect(component.textContent).toMatchInlineSnapshot(`"jscrazy"`);
			});

			it('should show "Former user" if no nickname', () => {
				const { getByTestId } = renderComponent({
					status: 'closed',
					nickname: undefined,
				});
				const component = getByTestId('profilecard-name');
				expect(component.textContent).toMatchInlineSnapshot(`"Former user"`);
			});
		});
	});

	describe('Disabled account message', () => {
		describe.each(['inactive', 'closed'] as Props['status'][])(
			'%s user',
			(status: Props['status']) => {
				it('should show provided message', () => {
					const { getByText } = renderComponent({
						status,
						disabledAccountMessage: 'This account is disabled',
					});
					expect(getByText('This account is disabled')).toBeDefined();
				});

				it(`should show "just now" message for ${status} user`, () => {
					const { getByTestId } = renderComponent({
						status,
						statusModifiedDate: Date.now() / 1000,
					});

					expect(getByTestId('profilecard-disabled-account').textContent).toEqual(
						`You can no longer collaborate with this person. Their account was ${
							status === 'inactive' ? 'deactivated' : 'deleted'
						} this week.`,
					);
				});

				it(`should show message for ${status} user when modified date is not available`, () => {
					const { getByTestId } = renderComponent({
						status,
						statusModifiedDate: undefined,
					});

					expect(getByTestId('profilecard-disabled-account').textContent).toEqual(
						`You can no longer collaborate with this person. Their account has been ${
							status === 'inactive' ? 'deactivated' : 'deleted'
						}.`,
					);
				});

				it('should show lozenge if enabled', () => {
					const { getByText } = renderComponent({
						status,
						hasDisabledAccountLozenge: true,
					});

					expect(
						getByText(`Account ${status === 'closed' ? 'deleted' : 'deactivated'}`),
					).toBeDefined();
				});

				it('should not show lozenge if disabled', () => {
					const { queryByText } = renderComponent({
						status,
						hasDisabledAccountLozenge: false,
					});

					expect(
						queryByText(`Account ${status === 'closed' ? 'deleted' : 'deactivated'}`),
					).toBeNull();
				});

				it('should render name as a heading', () => {
					(fg as jest.Mock).mockReturnValue(true);
					const { getByRole } = renderComponent({
						fullName: 'Same name',
						nickname: 'Same name',
					});
					const component = getByRole('heading', { level: 2 });
					expect(component).toBeVisible();
					expect(component).toHaveTextContent('Same name');
				});
			},
		);
	});

	describe('Custom lozenges', () => {
		const customLozenges: LozengeProps[] = [
			{
				text: 'Guest',
				appearance: 'new',
				isBold: true,
			},
			{
				text: 'Cool Bean',
				appearance: 'removed',
			},
			{
				text: <div>Another Role</div>,
				appearance: 'inprogress',
				isBold: true,
			},
		];

		const getTextOf = (elem: React.ReactElement | string | null): string => {
			if (typeof elem === 'string') {
				return elem.toString();
			}

			if (!elem) {
				return '';
			}

			return getTextOf(elem.props.children);
		};

		it('should render multiple lozenges as expected', () => {
			const { getByText } = renderComponent({
				customLozenges,
			});

			for (const { text } of customLozenges) {
				const childText = getTextOf(text as React.ReactElement | string);

				expect(getByText(childText)).not.toBeNull();
			}
		});

		it('should render only one lozenge if only one provided', () => {
			const { getByText } = renderComponent({
				customLozenges: [customLozenges[0]],
			});

			const childText = getTextOf(customLozenges[0].text as React.ReactElement | string);

			expect(getByText(childText)).not.toBeNull();
		});
	});
});
