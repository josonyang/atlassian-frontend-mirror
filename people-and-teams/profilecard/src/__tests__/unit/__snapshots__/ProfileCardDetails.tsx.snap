// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileCardDetails reportingLines property should match snapshot when no reporting lines are specified 1`] = `
<Styled(div)>
  <Styled(h2)
    data-testid="profilecard-name"
    id="profilecard-name-label"
    noMeta={true}
  >
    full name test (jscrazy) 
  </Styled(h2)>
  <CustomLozenges />
  <IconLabel
    icon="email"
  />
  <IconLabel
    icon="time"
  />
  <IconLabel
    icon="companyName"
  >
    Atlassian
  </IconLabel>
  <IconLabel
    icon="location"
  />
  <ReportingLinesDetails
    fireAnalyticsWithDuration={[MockFunction]}
  />
</Styled(div)>
`;

exports[`ProfileCardDetails reportingLines property should match snapshot when no reporting lines are specified 2`] = `
<Styled(div)>
  <Styled(h2)
    data-testid="profilecard-name"
    id="profilecard-name-label"
    noMeta={true}
  >
    full name test (jscrazy) 
  </Styled(h2)>
  <CustomLozenges />
  <IconLabel
    icon="email"
  />
  <IconLabel
    icon="time"
  />
  <IconLabel
    icon="companyName"
  >
    Atlassian
  </IconLabel>
  <IconLabel
    icon="location"
  />
  <ReportingLinesDetails
    fireAnalyticsWithDuration={[MockFunction]}
    reportingLines={{}}
  />
</Styled(div)>
`;

exports[`ProfileCardDetails reportingLines property should match snapshot when no reporting lines are specified 3`] = `
<Styled(div)>
  <Styled(h2)
    data-testid="profilecard-name"
    id="profilecard-name-label"
    noMeta={true}
  >
    full name test (jscrazy) 
  </Styled(h2)>
  <CustomLozenges />
  <IconLabel
    icon="email"
  />
  <IconLabel
    icon="time"
  />
  <IconLabel
    icon="companyName"
  >
    Atlassian
  </IconLabel>
  <IconLabel
    icon="location"
  />
  <ReportingLinesDetails
    fireAnalyticsWithDuration={[MockFunction]}
    reportingLines={
      {
        "managers": [],
        "reports": [],
      }
    }
  />
</Styled(div)>
`;

exports[`ProfileCardDetails reportingLines property should match snapshot when reporting lines are specified 1`] = `
<Styled(div)>
  <Styled(h2)
    data-testid="profilecard-name"
    id="profilecard-name-label"
    noMeta={true}
  >
    full name test (jscrazy) 
  </Styled(h2)>
  <CustomLozenges />
  <IconLabel
    icon="email"
  />
  <IconLabel
    icon="time"
  />
  <IconLabel
    icon="companyName"
  >
    Atlassian
  </IconLabel>
  <IconLabel
    icon="location"
  />
  <ReportingLinesDetails
    fireAnalyticsWithDuration={[MockFunction]}
    reportingLines={
      {
        "managers": [
          {
            "accountIdentifier": "abcd",
            "identifierType": "ATLASSIAN_ID",
            "pii": {
              "name": "name",
              "picture": "picture",
            },
          },
        ],
        "reports": [
          {
            "accountIdentifier": "abcd",
            "identifierType": "ATLASSIAN_ID",
            "pii": {
              "name": "name",
              "picture": "picture",
            },
          },
          {
            "accountIdentifier": "abcd",
            "identifierType": "ATLASSIAN_ID",
            "pii": {
              "name": "name",
              "picture": "picture",
            },
          },
        ],
      }
    }
  />
</Styled(div)>
`;
